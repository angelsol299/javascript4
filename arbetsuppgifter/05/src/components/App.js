import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { loadRates, mapObjectToArray } from '../api';
import Button from './Button';

export default class App extends Component {
  state = {
    rates: [],
    error: '',
    search: ''
  };

  onChange = ({ target }) => this.setState({ [target.name] : target.value });

  updateRates = () => {
    loadRates(this.props.base)
      .then(({ rates, base }) => {
        this.setState({ rates: mapObjectToArray(rates), base })
      })
      .catch(e => this.setState({ error: e.message }));
  };

  filterList = rate => {
    if(rate.value && this.state.search) {
      return rate.key.toLocaleLowerCase().includes(this.state.search.toLocaleLowerCase())
    }
    return rate;
  }

  renderList = rates => rates
    .filter(this.filterList)
    .map(rate => (
      <li className="py-2 px-4 mb-4 border rounded shadow" key={rate.key}>
        {rate.value} {rate.key}
      </li>
   ));

  render() {
    return (
      <div className="flex flex-col justify-center items-center w-1/2 mx-auto font-mono text-grey-darker mt-6">
        <div className="fixed pin-t pin-r m-4 font-mono uppercase text-sm flex flex-col items-center">
          <Button onClick={this.updateRates} />
        </div>
        <h1 className="text-pink-dark mb-4">{this.props.base} Rates</h1>
        <p className="error">{this.state.error}</p>
        <input 
          type="text" 
          value={this.state.search} 
          onChange={this.onChange} 
          name="search" 
          className="shadow appearance-none border rounded w-full py-2 px-3 mb-4 text-grey-darker" 
        />
        <ul className="list-reset">
          {this.renderList(this.state.rates)}
        </ul>
      </div>
    );
  }
}

App.propTypes = {
  base: PropTypes.string
};

/* 
 * If no props are sent down from `index.js` the component
 * uses these props, default is EUR and todays rate
 */
App.defaultProps = {
  base: 'EUR'
};