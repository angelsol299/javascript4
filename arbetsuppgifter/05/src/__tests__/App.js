import React from 'react';
import { shallow, mount, render } from 'enzyme';
import App from '../components/App';
import { mapObjectToArray, loadRates } from '../api';

test('should be base EUR', () => {
  const wrapper = shallow(<App />);
  expect(wrapper.find('h1').text()).toContain('EUR');
});

it('should match snapshot', () => {
  const wrapper = render(<App />);
  expect(wrapper).toMatchSnapshot();
});
