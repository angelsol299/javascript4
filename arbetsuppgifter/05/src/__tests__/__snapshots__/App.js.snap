// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  class="flex flex-col justify-center items-center w-1/2 mx-auto font-mono text-grey-darker mt-6"
>
  <div
    class="fixed pin-t pin-r m-4 font-mono uppercase text-sm flex flex-col items-center"
  >
    <button
      class="bg-pink hover:bg-pink-light 
  text-white 
  font-bold py-2 px-4 
  border-b-4 
  border-pink-dark 
  hover:border-pink rounded mt-4
"
      data-test="button"
    >
      Update Rates
    </button>
  </div>
  <h1
    class="text-pink-dark mb-4"
  >
    EUR Rates
  </h1>
  <p
    class="error"
  />
  <input
    class="shadow appearance-none border rounded w-full py-2 px-3 mb-4 text-grey-darker"
    date-test="input"
    name="search"
    type="text"
    value=""
  />
  <div
    data-test="list"
  />
</div>
`;
